crates/ruff_linter/src/noqa.rs

```diff
@@ -524,26 +546,127 @@ fn add_noqa_inner(
     noqa_line_for: &NoqaMapping,
     line_ending: LineEnding,
 ) -> (usize, String) {
-    // Map of line start offset to set of (non-ignored) diagnostic codes that are triggered on that line.
-    let mut matches_by_line: BTreeMap<TextSize, (RuleSet, Option<&Directive>)> =
-        BTreeMap::default();
+    let mut count = 0;
 
     // Whether the file is exempted from all checks.
     // Codes that are globally exempted (within the current file).
     let exemption =
         FileExemption::try_extract(locator.contents(), comment_ranges, external, path, locator);
     let directives = NoqaDirectives::from_commented_ranges(comment_ranges, path, locator);
 
+    let comments = find_noqa_comments(diagnostics, locator, &exemption, &directives, noqa_line_for);
+
+    let edits = build_noqa_edits_by_line(comments, locator, line_ending);
+
+    let contents = locator.contents();
+
+    let mut output = String::with_capacity(contents.len());
+    let mut last_append = TextSize::default();
+
+    for (_, edit) in edits {
+        output.push_str(&contents[TextRange::new(last_append, edit.start())]);
+
+        edit.write(&mut output);
+
+        count += 1;
+
+        last_append = edit.end();
+    }
+
+    output.push_str(&contents[TextRange::new(last_append, TextSize::of(contents))]);
+
+    (count, output)
+}
+
+fn build_noqa_edits_by_diagnostic(
+    comments: Vec<Option<NoqaComment>>,
+    diagnostics: &[Diagnostic],
+    locator: &Locator,
+    line_ending: LineEnding,
+) -> Vec<Option<Edit>> {
+    let mut edits = Vec::default();
+    for (i, comment) in comments.into_iter().enumerate() {
+        match comment {
+            Some(comment) => {
+                debug_assert_eq!(comment.diagnostic, &diagnostics[i]);
```

Comment:
Should we zip here instead? (I suppose the diagnostics could be omitted entirely?)